lib/compiler-singlepass/src/codegen_x64.rs

```diff
@@ -1924,14 +1925,44 @@ impl<'a> FuncGen<'a> {
         id
     }
 
+    fn emit_stack_check(&mut self, enter: bool) {
+        // `local_types` include parameters as well.
+        let locals = self.local_types.len()
+            // we add 1 to ensure that deep recursion is prohibited even for local and argument free
+            // functions, as they still use stack space for the return address.
+            + 1;
```

Comment:
Are we sure that `+1` is the best constant here?

Reply:
It isn't a great constant, but I believe this matches [the intent of the spec](https://webassembly.github.io/spec/core/exec/runtime.html#syntax-frame) the best. In particular I believe what we're trying to account for here are the number of entries (of any kind) on the unified wasm stack.

In practice it definitely seems like some of the entries may use more of the host stack than the others. I believe the solution to that, however, is in allocating the host stack in such a way that it can accommodate for a worst case wasm stack.

Reply:
Yeah, this sounds reasonable. One bit of doubt I still have is that function activation frame seems like a potentially obviously bigger entity than anything else.

Reply:
Yeah, x86_64 requires call stack to be aligned to 16 bytes, so each function call is going to be at least an equivalent of the `v128` value entry.

Reply:
Let's make it 4 at least, as:
   * we store return address and frame pointer
   * we also typically store couple callee saved regs to use as context register and as register allocator budget
  In practise, it's way bigger, as Wasmer implementation also allocates 32-byte red zone and few other parts.
Anyway, let's tune constants a bit.

Reply:
My worry is that changing the constants here is going to be a protocol breaking change by itself in the future, although probably not one which we'd be too worried about being observable.

The way I see it, we'd ideally change how we think about the size of a single stack frame rather than trying to introduce some sort of a multiplier for the size of some specific stack frame type. That is, rather than saying a frame is basically 8 bytes and then making one of the frames occupy 4 “slots”, we'd say that any frame is going to be at most 32 bytes, 64 bytes or somesuch (enforcing this limit in codegen).

Such approach at the very least would allow us to easily allocate a host stack according to the WASM stack limits. Is it somewhat wasteful? Yeah, but probably not to the point where the waste would be excessive. Easy to understand and verify? I think it is potentially much better than alternatives.

Reply:
I guess our actual limits shall be enforced by instrumentation, and indeed this code says that we presume frame at least uses 32 bytes